pkg/provider-local/controller/extension/shoot/actuator.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shoot
+
+import (
+	"context"
+	"time"
+
+	"github.com/gardener/gardener/extensions/pkg/controller/extension"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	gardenerkubernetes "github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ApplicationName is the name of the application.
+	ApplicationName string = "local-ext-shoot"
+	// ManagedResourceNamesShoot is the name used to describe the managed shoot resources.
+	ManagedResourceNamesShoot string = ApplicationName
+)
+
+type actuator struct {
+	client client.Client
+}
+
+// NewActuator returns an actuator responsible for Extension resources.
+func NewActuator() extension.Actuator {
+	return &actuator{}
+}
+
+// InjectClient injects the controller runtime client into the reconciler.
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+// Reconcile the extension resource.
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	namespace := ex.Namespace
+
+	shootResources, err := getShootResources()
+	if err != nil {
+		return err
+	}
+
+	var (
+		injectedLabels       = map[string]string{v1beta1constants.ShootNoCleanup: "true"}
+		secretNameWithPrefix = true
+		keepObjects          = false
+	)
+
+	return managedresources.Create(
```

Comment:
Is there any reason to call `managedresources.Create` instead of `managedresources.CreateForShoot` here? The latter one should be a bit easier to understand since we don't have to pass so many parameters.

Reply:
Yes there is. `CreateForShoot` uses `NewForShoot` and the later adds the `origin=gardener` label to the managed resources. Then during shoot deletion Gardener deletes all managed resources with that label and deletes the mr before the extension gets a chance to delete it on its own. This is actually a bug right now which is present in some of the extensions (all that use this function) but it is not noticed since the deletion of the extension resources is right after the mentioned step see [here](https://github.com/gardener/gardener/blob/06b51b25c18c52176fcda56a5677cc5e87efaece/pkg/gardenlet/controller/shoot/shoot_control_delete.go#L418). I plan to address this by changing the CreateForShoot function to not always add this label, but this is a change that does not concern this PR

Reply:
I see, I didn't know that there was an issue related to the `origin=gardener` label. Thanks for the explanation!

Reply:
I think that there is no need for a TODO. This whole new logic with the `BeforeKubeAPIServer` will not work as expected with this function. I can open an issue if you want

Reply:
Here is a link to the issue https://github.com/gardener/gardener/issues/7141